src/akimbo/io.py

```diff
@@ -24,44 +44,124 @@ def read_parquet(
     Parameters
     ----------
     url: data location
+        Directory with data files, single file or glob pattern
     storage_options: any arguments for an fsspec backend
     extract: whether to turn top-level records into a dataframe. If False,
         will return a series.
+    backend: one of "pandas", "polars" or "dask"
     """
+    # TODO: is this useful compared to pyarrow.parquet? Describe differences
     ds = ak.from_parquet(url, storage_options=storage_options, **kwargs)
-    s = akimbo.pandas.PandasAwkwardAccessor._to_output(ds)
-    if extract:
-        return s.ak.unmerge()
-    return s
+    return ak_to_series(ds, backend, extract=extract)
 
 
 def read_json(
-    url,
+    url: str,
     storage_options: dict | None = None,
-    extract=True,
+    extract: bool = True,
+    backend: str = "pandas",
     **kwargs,
 ):
-    """Read a JSON dataset with nested data into a pandas Series or DataFrame.
+    """Read a JSON dataset with nested data into a Series or DataFrame.
 
     You can pass a selection of columns to read (list or jsonschema format),
     using ``schema=``, and
     other columns will not be parsed into memory. See the docs for
     ak.from_json for further details.
 
+    (examples to come)
+
     Parameters
     ----------
-    url: data location
+    url: data location (may include glob characters)
     storage_options: any arguments for an fsspec backend
     extract: whether to turn top-level records into a dataframe. If False,
         will return a series.
+    backend: one of "pandas", "polars" or "dask"
     """
-    with fsspec.open(url, **storage_options) as f:
-        ds = ak.from_json(
-            f,
-            line_delimited=True,
-            **kwargs,
-        )
-    s = akimbo.pandas.PandasAwkwardAccessor._to_output(ds)
-    if extract:
-        return s.ak.unmerge()
-    return s
+    with fsspec.open_files(url, **(storage_options or {})) as f:
+        ds = ak.concatenate([ak.from_json(_, line_delimited=True, **kwargs) for _ in f])
+    return ak_to_series(ds, backend, extrcact=extract)
+
+
+def read_avro(
+    url: str,
+    storage_options: dict | None = None,
+    extract: bool = True,
+    backend: str = "pandas",
+    **kwargs,
+):
+    """Read AVRO structured data files
+
+    Parameters
+    ----------
+    url: data location (may include glob characters)
+    storage_options: any arguments for an fsspec backend
+    extract: whether to turn top-level records into a dataframe. If False,
+        will return a series.
+    backend: one of "pandas", "polars" or "dask"
+    """
+    with fsspec.open(url, **(storage_options or {})) as f:
+        ds = ak.from_avro_file(f, **kwargs)
+    return ak_to_series(ds, backend, extract=extract)
+
+
+def _merge(ind1, ind2, builder):
```

Comment:
@jpivarski , is it expected that including a builder should make this not cacheable by njit? 

I was thinking about making an offsets array, but we don't know how big the values array will be, as some IDs in ind2 may not exist in ind1 - we only have an upper bound. Perhaps a starts/stops pair would be fine for this.

Reply:
I've never tested cacheability of Numba functions to take any kind of Awkward arguments. (You're not talking about a closure, right? Numba closures have to include the objects as constants.) In principle, there's no reason why a function with a ArrayBuilder argument couldn't be cached, unless maybe it has some associated behavior that is not serializable. Behavior dicts are part of an Awkward argument's type.

Reply:
In the only usage I have, the builder is pristine at the start of the function. For ak arrays I've passed haven't had any attributes of behaviours. The specific message is
> NumbaWarning: Cannot cache compiled function "_merge" as it uses dynamic globals (such as ctypes pointers and large global arrays)